launchpod/angular-launchpod/backend/src/main/java/com/google/launchpod/servlets/FormHandlerServlet.java

```diff
@@ -0,0 +1,294 @@
+  }
+
+  /**
+   * Generate policy for directly uploading a file to Cloud Storage via HTML form.
+   * @return HTML form (as a String) for uploading MP3
+   */
+  public String generateSignedPostPolicyV4 (String projectId, String bucketName, String blobName) throws IOException {
+    // String blobName = entity ID from Datastore, as the name of object uploaded to GCS
+
+    String errorMessage = "Policy is null.";
+    Storage storage;
+    PostPolicyV4.PostFieldsV4 fields;
+    PostPolicyV4 policy = null;
+    String myRedirectUrl = "";
+
+    try {
+      // Prepare credentials and API key
+      String keyFileName = "launchpod-step18-2020-47434aafba88.json";
+      ClassLoader classLoader = getClass().getClassLoader();
+      File file = new File(classLoader.getResource(keyFileName).getFile());
+      GoogleCredentials credentials = GoogleCredentials.fromStream(new FileInputStream(file));
+
+      storage = StorageOptions.newBuilder().setCredentials(credentials).build().getService();
+
+      myRedirectUrl = GENERATE_RSS_URL + blobName;
+      fields =
+          PostPolicyV4.PostFieldsV4.newBuilder().setSuccessActionRedirect(myRedirectUrl).build();
+
+      policy =
+        storage.generateSignedPostPolicyV4(
+            BlobInfo.newBuilder(bucketName, blobName).build(), 10, TimeUnit.MINUTES, fields);
+    } catch(Exception e) {
+      errorMessage = e.getMessage();
+      throw e;
+    }
+
+    // StringBuilder htmlForm;
+    StringBuilder htmlForm = new StringBuilder();
+    if (policy!=null) {
+      // htmlForm =
+      //     new StringBuilder(
+      //         "<form name='mp3-upload' action='"
+      //             + policy.getUrl()
+      //             + "' method='POST' enctype='multipart/form-data'>\n");
+      String htmlFormPt1 = String.format("<form name='mp3-upload' action='%s' method='POST' enctype='multipart/form-data'>\n", policy.getUrl());
+      htmlForm.append(htmlFormPt1);
+
+      for (Map.Entry<String, String> entry : policy.getFields().entrySet()) {
+        String htmlFormPt2 = String.format("<input name='%s' value='%s' type='hidden' />\n", entry.getKey(), entry.getValue());
+        htmlForm.append(htmlFormPt2);
+      }
+      htmlForm.append(HTML_FORM_END);
+    } else {
+      htmlForm = new StringBuilder();
+      htmlForm.append(errorMessage);
+    }
+    return htmlForm.toString();
+  }
+
+  /**
+  * Display RSS feed xml string that user tries recalling with the given ID.
+  */
+  @Override
+  public void doGet(HttpServletRequest req, HttpServletResponse res) throws IOException {
+    String actionString = req.getParameter(ACTION);
+    String id = req.getParameter(ID);
+
+    if (actionString==null || id==null) {
+      res.setContentType("text/html");
+      res.getWriter().println("Please specify action and/or id.");
+      res.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+      return;
+    }
+
+    Action act = Action.generateRSSLink; // initialization
+    Action action = act.valueOf(actionString);
```

Comment:
this throws an NPE Or IAE for incorrect values - you need to catch those issues: https://docs.oracle.com/javase/7/docs/api/java/lang/Enum.html#valueOf(java.lang.Class,%20java.lang.String). 

alternatively, you could just change these if statements to check against the string for each enum value directly, especially since you have an else at the end of the block

Reply:
I'll catch an IAE - I don't think it would throw an NPE because I previously check for a null action above